jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1128 @@
+         *
+         * @return a  <code>Spliterator</code> over the elements in the backing set's unmodifiable copy.
+         */
+        @Override
+        public Spliterator<E> spliterator()
+        {
+            return getCopy().spliterator();
+        }
+
+        /**
+         * Return a sequential <code>Stream</code> with the backing set's unmodifiable copy as its
+         * source.
+         * @return a sequential <code>Stream</code> with the backing set's unmodifiable copy as its
+         * source.
+         */
+        @Override
+        public Stream<E> stream()
+        {
+            return getCopy().stream();
+        }
+
+        /**
+         * Return a possibly parallel <code>Stream</code> with the backing set's unmodifiable copy as
+         * its source.
+         * @return a possibly parallel <code>Stream</code> with the backing set's unmodifiable copy
+         * as its source.
+         * It is allowable for this method to return a sequential stream.
+         */
+        @Override
+        public Stream<E> parallelStream()
+        {
+            return getCopy().parallelStream();
+        }
+
+        /**
+         * Compares the specified object with this set for equality.
+         * @param o object to be compared for equality with this set.
+         * @return <code>true</code> if o and this set are the same object or o is equal to the
+         * backing object, false otherwise.
+         */
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o)
+                return true;
+            synchronized (mutex) {
+                return set.equals(o);
+            }
+        }
+
+        /**
+         * Return the backing set's hashcode.
+         * @return the backing set's hashcode.
+         */
+        @Override
+        public int hashCode()
+        {
+            synchronized (mutex) {
+                return set.hashCode();
+            }
+        }
+
+        /**
+         * Return the backing set's toString result.
+         * @return the backing set's toString result.
+         */
+        @Override
+        public String toString()
+        {
+            synchronized (mutex) {
+                return set.toString();
+            }
+        }
+
+        /**
+         * Get the backing set's unmodifiable copy.
+         * @return the backing set's unmodifiable copy.
+         */
+        private Set<E> getCopy()
+        {
+            synchronized (mutex) {
+                if (copy != null)
+                    return copy;
+                return Collections.unmodifiableSet(new LinkedHashSet<>(set));
+            }
+        }
+
+        /**
+         * If the backing set is modified, call this method to let this set knows the backing set's
+         * copy need to update.
+         */
+        private void modified()
+        {
+            synchronized (mutex) {
+                copy = null;
+            }
+        }
```

Comment:
It is not safe if this method is not synchronized.

Reply:
It's actually safe since a single mutex is shared with the outer class, and all of the calling methods in the outer class are synchronized.  But there's no harm in double synchronization in this case.

Reply:
Oh, I am wrong about it. Thanks